File: services/spoolPrediction.ts
import { Tier, Bracket, RankedItem, PredictionSignals } from '../types';
import { TIER_SCORE_RANGES, NEW_USER_THRESHOLD } from '../constants';
import { computeTierScore } from './rankingAlgorithm';

const WEIGHTS = {
  genreAffinity: 0.45,
  globalScore: 0.35,
  bracketAffinity: 0.20,
} as const;

export function computePredictionSignals(
  allItems: RankedItem[],
  primaryGenre: string,
  bracket: Bracket,
  globalScore: number | undefined,
  tier: Tier,
): PredictionSignals {
  const range = TIER_SCORE_RANGES[tier];

  // Genre affinity: average score of user's movies with same primary genre
  const genreItems = allItems.filter(
    (item) => item.genres.length > 0 && item.genres[0] === primaryGenre
  );
  let genreAffinity: number | null = null;
  if (genreItems.length > 0) {
    const scores = genreItems.map((item) => {
      const tierRange = TIER_SCORE_RANGES[item.tier];
      const tierPeers = allItems
        .filter((i) => i.tier === item.tier)
        .sort((a, b) => a.rank - b.rank);
      return computeTierScore(item.rank, tierPeers.length, tierRange.min, tierRange.max);
    });
    genreAffinity = scores.reduce((a, b) => a + b, 0) / scores.length;
  }

  // Global score: clamp to tier range
  let mappedGlobal: number | null = null;
  if (globalScore !== undefined) {
    mappedGlobal = Math.max(range.min, Math.min(range.max, globalScore));
  }

  // Bracket affinity: average score of user's movies with same bracket
  const bracketItems = allItems.filter((item) => item.bracket === bracket);
  let bracketAffinity: number | null = null;
  if (bracketItems.length > 0) {
    const scores = bracketItems.map((item) => {
      const tierRange = TIER_SCORE_RANGES[item.tier];
      const tierPeers = allItems
        .filter((i) => i.tier === item.tier)
        .sort((a, b) => a.rank - b.rank);
      return computeTierScore(item.rank, tierPeers.length, tierRange.min, tierRange.max);
    });
    bracketAffinity = scores.reduce((a, b) => a + b, 0) / scores.length;
  }

  return {
    genreAffinity,
    globalScore: mappedGlobal,
    bracketAffinity,
    totalRanked: allItems.length,
  };
}

export function predictScore(signals: PredictionSignals, tier: Tier): number {
  const range = TIER_SCORE_RANGES[tier];
  const midpoint = (range.min + range.max) / 2;

  // New user fallback: use globalScore only
  if (signals.totalRanked < NEW_USER_THRESHOLD) {
    return signals.globalScore ?? midpoint;
  }

  // Build weighted average from available signals
  const entries: { value: number; weight: number }[] = [];

  if (signals.genreAffinity !== null) {
    entries.push({ value: signals.genreAffinity, weight: WEIGHTS.genreAffinity });
  }
  if (signals.globalScore !== null) {
    entries.push({ value: signals.globalScore, weight: WEIGHTS.globalScore });
  }
  if (signals.bracketAffinity !== null) {
    entries.push({ value: signals.bracketAffinity, weight: WEIGHTS.bracketAffinity });
  }

  if (entries.length === 0) return midpoint;

  // Redistribute weights proportionally
  const totalWeight = entries.reduce((sum, e) => sum + e.weight, 0);
  const raw = entries.reduce((sum, e) => sum + e.value * (e.weight / totalWeight), 0);

  // Clamp to tier bounds
  return Math.max(range.min, Math.min(range.max, Math.round(raw * 100) / 100));
}

File: ios/Spool/Sources/Spool/Algorithm/SpoolPrediction.swift
import Foundation

/// Pure port of `services/spoolPrediction.ts`.
/// Signal-based score prediction (genre + global + bracket).
public enum SpoolPrediction {

    // Internal (not private) so @testable parity tests can assert these
    // against fixtures/engine-parity.json predictionWeights.
    static let weightGenreAffinity:   Double = 0.45
    static let weightGlobalScore:     Double = 0.35
    static let weightBracketAffinity: Double = 0.20

    public static func computePredictionSignals(
        allItems: [RankedItem],
        primaryGenre: String,
        bracket: Bracket,
        globalScore: Double?,
        tier: Tier
    ) -> PredictionSignals {
        let range = tier.scoreRange

        // Genre affinity: average score of items with same primary genre
        let genreItems = allItems.filter { !$0.genres.isEmpty && $0.genres[0] == primaryGenre }
        let genreAffinity: Double? = averageScore(for: genreItems, in: allItems)

        // Global score: clamp to tier range
        let mappedGlobal: Double? = globalScore.map { max(range.min, min(range.max, $0)) }

        // Bracket affinity: average score of items with same bracket
        let bracketItems = allItems.filter { $0.bracket == bracket }
        let bracketAffinity: Double? = averageScore(for: bracketItems, in: allItems)

        return PredictionSignals(
            genreAffinity: genreAffinity,
            globalScore: mappedGlobal,
            bracketAffinity: bracketAffinity,
            totalRanked: allItems.count
        )
    }

    public static func predictScore(signals: PredictionSignals, tier: Tier) -> Double {
        let range = tier.scoreRange
        let midpoint = (range.min + range.max) / 2

        if signals.totalRanked < SpoolConstants.newUserThreshold {
            return signals.globalScore ?? midpoint
        }

        var entries: [(value: Double, weight: Double)] = []
        if let v = signals.genreAffinity   { entries.append((v, weightGenreAffinity)) }
        if let v = signals.globalScore     { entries.append((v, weightGlobalScore)) }
        if let v = signals.bracketAffinity { entries.append((v, weightBracketAffinity)) }
        if entries.isEmpty { return midpoint }

        let totalWeight = entries.reduce(0) { $0 + $1.weight }
        let raw = entries.reduce(0) { $0 + $1.value * ($1.weight / totalWeight) }
        let rounded = (raw * 100).rounded() / 100
        return max(range.min, min(range.max, rounded))
    }

    // MARK: helper

    private static func averageScore(for items: [RankedItem], in allItems: [RankedItem]) -> Double? {
        guard !items.isEmpty else { return nil }
        let scores: [Double] = items.map { item in
            let tierRange = item.tier.scoreRange
            let tierPeers = allItems.filter { $0.tier == item.tier }.sorted { $0.rank < $1.rank }
            let positionInTier = tierPeers.firstIndex(where: { $0.id == item.id }) ?? 0
            return RankingAlgorithm.computeTierScore(
                position: positionInTier, totalInTier: tierPeers.count,
                tierMin: tierRange.min, tierMax: tierRange.max
            )
        }
        return scores.reduce(0, +) / Double(scores.count)
    }
}
